dataflow/src/main/java/org/checkerframework/dataflow/util/UnmodifiableIdentityHashMap.java

```diff
@@ -0,0 +1,195 @@
+ * underlying map.
+ *
+ * <p>This class extends {@link IdentityHashMap} only so it is assignable to variables / fields of
+ * static type {@link IdentityHashMap}. All valid operations are delegated to the wrapped map, and
+ * any inherited state from the superclass is unused.
+ */
+@SuppressWarnings("keyfor")
+public class UnmodifiableIdentityHashMap<K, V> extends IdentityHashMap<K, V> {
+
+  private static final long serialVersionUID = -5147442142854693854L;
+
+  /** the wrapped map */
+  private final IdentityHashMap<K, V> map;
+
+  /**
+   * Create an UnmodifiableIdentityHashMap
+   *
+   * @param map the map to wrap
+   */
+  private UnmodifiableIdentityHashMap(IdentityHashMap<K, V> map) {
+    this.map = map;
+  }
+
+  /**
+   * Create an {@link UnmodifiableIdentityHashMap} wrapper for a map.
+   *
+   * @param map the map to wrap
+   * @return the wrapper
+   * @param <K> the key type
+   * @param <V> the value type
+   */
+  public static <K, V> UnmodifiableIdentityHashMap<K, V> wrap(IdentityHashMap<K, V> map) {
+    // avoid repeated wrapping
+    if (map instanceof UnmodifiableIdentityHashMap) {
+      return (UnmodifiableIdentityHashMap<K, V>) map;
+    }
+    return new UnmodifiableIdentityHashMap<>(map);
+  }
+
+  @Override
+  public int size() {
+    return map.size();
+  }
+
+  @Override
+  public boolean isEmpty() {
+    return map.isEmpty();
+  }
+
+  @Override
+  public @Nullable V get(@Nullable Object key) {
+    return map.get(key);
+  }
+
+  @Override
+  public boolean containsKey(@Nullable Object key) {
+    return map.containsKey(key);
+  }
+
+  @Override
+  public boolean containsValue(@Nullable Object value) {
+    return map.containsValue(value);
+  }
+
+  @Override
+  public @Nullable V put(K key, V value) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void putAll(Map<? extends K, ? extends V> m) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public @Nullable V remove(@Nullable Object key) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void clear() {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  @SuppressWarnings("IdentityHashMapUsage")
+  public boolean equals(@Nullable Object o) {
+    return map.equals(o);
+  }
+
+  @Override
+  public int hashCode() {
+    return map.hashCode();
+  }
+
+  @Override
+  public Object clone() {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public Set<K> keySet() {
+    return Collections.unmodifiableSet(map.keySet());
+  }
+
+  @Override
+  public Collection<V> values() {
+    return Collections.unmodifiableCollection(map.values());
+  }
+
+  @Override
+  public Set<Map.Entry<K, V>> entrySet() {
+    return Collections.unmodifiableSet(map.entrySet());
+  }
+
+  @Override
+  public void forEach(BiConsumer<? super K, ? super V> action) {
+    map.forEach(action);
```

Comment:
What if `action` performs side effects?  Is it up to the client to ensure that it does not?

Reply:
I think `action` is only passed the keys and values contained in the map, but not a direct reference to the wrapped map; so it should not be able to perform unwanted side effects on the map.  Did I miss something?

Reply:
You are right.  Thanks for clarifying.